src/SFA.DAS.EAS.Portal/Application/Commands/Command.cs

```diff
@@ -0,0 +1,72 @@
+using System;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using SFA.DAS.EAS.Portal.Application.Services;
+using SFA.DAS.EAS.Portal.Client.Database.Models;
+using SFA.DAS.EAS.Portal.Client.Types;
+
+namespace SFA.DAS.EAS.Portal.Application.Commands
+{
+    public class Command
```

Comment:
This is poor design and will lead to spaghetti code.  We shouldn't have a base command trying to implement all this functionality e.g. GetOrAddProvider, GetOrAddCohort.  That is not the concern of a command.  A command is a lightweight object containing some data.

Reply:
We had this discussion in the 454 PR 😄! Here's my position from one of the relevant [comments ](https://github.com/SkillsFundingAgency/das-employerapprenticeshipsservice/pull/1632#discussion_r284327408)from that PR, so that we can continue the discussion...

> Stephek 20 days ago  Member
> Is this a command or command handler? Should be separation of concerns as per the standard pattern command.handler pattern.
> 
> lazcool 19 days ago  Member
> It's a command object and there is no 'standard' pattern.
> 
> From [wikipedia's page on the command pattern](https://en.wikipedia.org/wiki/Command_pattern):
> 'In some implementations the handler code is part of the command/action object. In other implementations the code is part of the Receiver/Target Object, and in yet other implementations the handler code is kept separate from the other objects.'
> 

We also discussed the Command vs Command+Command Handler pattern in another [comment](https://github.com/SkillsFundingAgency/das-employerapprenticeshipsservice/pull/1632#discussion_r286857670). My point from that comment was, essentially, that the pattern where we have just a Command class is simpler, requires less code and performs better.

If we accept that a command object can encapsulate a command on its own, then it makes sense to put common functionality in a base class.

Reply:
Base classes should be for inheritance not common code.

Reply:
Please see my other [comment](https://github.com/SkillsFundingAgency/das-employerapprenticeshipsservice/pull/1646#issuecomment-498708165), which details how this adhers to the Liskov Substitution Principle (re your comment about following SOLID).

Reply:
GetOrAddProvider, GetOrAddCohort is not common command code.  As per my original comment this code should not reside in a Command base object and therefore there is no need for a base object.